### PR TITLE
Add error for missing await in conditionals

### LINKED ISSUE #25330: Error for missing await in conditional
**TypeScript Version:**  3.0.0-dev.20180626

**Search Terms:** 
- await
- conditional
- diagnostics

**Code**

```ts
async function foo(): Promise<boolean> {
    return false;
}

async function bar(): Promise<string> {
    if (foo()) {
        return 'true';
    }
    return 'false';
}
```

**Expected behavior:**
A warning or error or suggestion is generated in the `if` conditional since it will always be true. The user almost certainly intended to await the result of `foo` instead.

Possibly also have a quick fix to add the await

**Actual behavior:**
No errors or warnings. Code does unexpected things

### PR DESCRIPTION
**Pull request by Jack-Works**
image

close #25330

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot test dt
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by Jack-Works** [comment]
https://github.com/typescript-bot/TypeScript/pull/49/files#diff-0e62d568d28b82cee9e66b2b8fafd290R7

Is this a possible bug in vscode? 🤔

**Comment by DanielRosenwasser** [comment]
@mjbvz if we can get some of the numbers down, we'd like to get this merged in for a future release. Like @Jack-Works mentioned, I think this check might've caught a potential bug in VS Code in an un-awaited promise.

**Comment by mjbvz** [comment]
Thanks @Jack-Works!

The error in `editorActions` looks like it should not cause problems. We can probably just remove the conditional

The error in `server` looks like bad typings since the value really can be undefined (but we don't type it that way)

**Comment by Jack-Works** [comment]
Another question, should it behave like the check to function? If the variable is undefined | Promise, and use as a if condition. If it is not used in the if body, should it be an error?

**Comment by Jack-Works** [comment]
How's going 👀

**Comment by Jack-Works** [comment]
rebased to master and merge conflict resolved

**Comment by Jack-Works** [comment]
rebased to master and merge conflict resolved

**Comment by Jack-Works** [comment]
rebased to master and conflict resolved

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -34527,7 +34529,18 @@ namespace ts {
             checkSourceElement(node.elseStatement);
         }
 
-        function checkTestingKnownTruthyCallableType(condExpr: Expression, type: Type, body: Statement | Expression | undefined) {
+        function checkTestingKnownTruthyAwaitableType(condExpr: Expression, type: Type) {
```
Comment (elibarzilay) [inline src/compiler/checker.ts:34532]: This is used *almost* in every place that `checkTestingKnownTruthyCallableType` is used:

* There's one place where `checkTestingKnownTruthyCallableType` is used (`checkBinaryExpression`) but you didn't add `checkTestingKnownTruthyAwaitableType` --- it looks like it should be used there too, no?

* If it should be used there too, then it would be simpler to combine the two functions: instead of adding `checkTestingKnownTruthyAwaitableType`, just rename the existing one something like `checkTestingKnownTruthyCallableOrAwaitableType`.

* It would also simplify the code a bit.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -34538,14 +34551,8 @@ namespace ts {
...
 
-            if (!testedNode) {
-                return;
-            }
+            if (!testedNode) return;
 
-            const possiblyFalsy = getFalsyFlags(type);
-            if (possiblyFalsy) {
-                return;
-            }
```
Comment (elibarzilay) [inline src/compiler/checker.ts:34548]: Why is this removed?

(Also nitpick: you left two empty lines.)

**Review by elibarzilay — CHANGES_REQUESTED** [review]
See the comments.

(Another pickiness: the title of the PR and the commit could be better, since it's not only in `if` statements.  The title of the linked issue is probably good for this PR too.)

**Review by elibarzilay — APPROVED** [review]

**Comment by elibarzilay** [comment]
Rebased, resolved conflicts, squashed; re-edited the PR subject and the commit message.
